web/src/reducers/dirty.js

```diff
@@ -0,0 +1,167 @@
+    default:
+      return prop;
+  }
+};
+
+const up = updates =>
+  typeof updates === 'object'
+    ? Object.entries(updates).reduce(
+        (a, [prop, value]) => ({
+          ...a,
+          [mapActivityPropToAPI(prop)]: up(value)
+        }),
+        {}
+      )
+    : true;
+
+const dirty = (state = initialState, action) => {
+  switch (action.type) {
+    case SAVE_APD_SUCCESS:
+    case SELECT_APD:
+      return initialState;
+
+    case ADD_ACTIVITY_DIRTY:
+      // when an activity is added, we need to add the
+      // whole new activity to the dirty state since it
+      // is populated with initial data
+      return u(
+        {
+          dirty: true,
+          data: { activities: { byKey: { [action.data.key]: action.data } } }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_CONTRACTOR:
+    case REMOVE_ACTIVITY_CONTRACTOR:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { contractorResources: true } }
+            }
+          }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_GOAL:
+    case REMOVE_ACTIVITY_GOAL:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { goals: true } }
+            }
+          }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_EXPENSE:
+    case REMOVE_ACTIVITY_EXPENSE:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { expenses: true } }
+            }
+          }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_MILESTONE:
+    case REMOVE_ACTIVITY_MILESTONE:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { schedule: true } }
+            }
+          }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_STATE_PERSON:
+    case REMOVE_ACTIVITY_STATE_PERSON:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { statePersonnel: true } }
+            }
+          }
+        },
+        state
+      );
+
+    case REMOVE_ACTIVITY:
+      return u({ dirty: true }, state);
+
+    case UPDATE_ACTIVITY:
+      return u(
+        {
+          dirty: true,
+          data: { activities: { byKey: { [action.key]: up(action.updates) } } }
+        },
+        state
+      );
+
+    case UPDATE_APD:
+      if (
+        action.updates.federalCitations ||
+        action.updates.incentivePayments ||
+        action.updates.narrativeHIE ||
+        action.updates.narrativeHIT ||
+        action.updates.narrativeMMIS ||
+        action.updates.programOverview ||
+        action.updates.previousActivityExpenses ||
+        action.updates.previousActivitySummary ||
+        action.updates.stateProfile
+      ) {
```

Comment:
These are the APD fields that are filtered on save, so we need to explicitly mark them as dirty when they're changed.